tensorflow/compiler/tf2xla/kernels/bincount_op.cc

```diff
@@ -0,0 +1,126 @@
+    auto weights_shape = weights_shape_or.ValueOrDie();
+    auto weights_size = weights_shape.dimensions(0);
+    auto input_xla_type = ctx->input_xla_type(0);
+    xla::PrimitiveType dtype;
+    bool has_weights;
+    if (weights_size){
+      has_weights = true;
+      dtype = ctx->input_xla_type(2);
+    }
+    else {
+      has_weights = false;
+      dtype = input_xla_type;
+    }
+    int64_t output_size;
+    ctx->ConstantInputAsIntScalar("size", &output_size);
+    StatusOr<xla::Shape> input_shape_or = ctx->builder()->GetShape(input);
+    OP_REQUIRES_OK(ctx, input_shape_or.status());
+    auto input_shape = input_shape_or.ValueOrDie();
+    auto size = input_shape.dimensions(0);
+    auto rank = input_shape.rank();
+
+    xla::Shape output_shape = xla::ShapeUtil::MakeShape(dtype, {output_size});
+    xla::XlaOp idx, updates, output;
+    xla::ScatterDimensionNumbers scatter_dnums;
+    scatter_dnums.set_index_vector_dim(1);
+    scatter_dnums.add_inserted_window_dims(0);
+    scatter_dnums.add_scatter_dims_to_operand_dims(0);    
+    auto one = xla::One(ctx->builder(), input_xla_type);
+    auto zero = xla::Zero(ctx->builder(), input_xla_type);;
+    
+    if (rank == 2) {
+      output_shape = xla::ShapeUtil::MakeShape(dtype, {size, output_size});
+      scatter_dnums.add_inserted_window_dims(1);
+      scatter_dnums.add_scatter_dims_to_operand_dims(1);
+      auto i_shape = xla::ShapeUtil::MakeShape(input_xla_type, {input_shape.dimensions()});
+      auto i = xla::Iota(ctx->builder(), i_shape, 0);
+      i = xla::Reshape(i, {input_shape.dimensions(0)*input_shape.dimensions(1), 1});
+      auto j = xla::Reshape(input, {input_shape.dimensions(0)*input_shape.dimensions(1), 1});
+      std::vector<xla::XlaOp> iotas_to_concat;
+      iotas_to_concat.push_back(i);
+      iotas_to_concat.push_back(j);
+      idx = xla::ConcatInDim(ctx->builder(), iotas_to_concat, 1);
+      updates = xla::Broadcast(one, {input_shape.dimensions(0)*input_shape.dimensions(1)});
+      if (has_weights) {
+        weights = xla::Reshape(weights, {input_shape.dimensions(0)*input_shape.dimensions(1)});
+        zero = xla::Zero(ctx->builder(), dtype);
+        updates = weights;
+      }
+    } else {
+      input = xla::Reshape(input, {size, 1});
+      idx = xla::Reshape(input, {size, 1});
+      updates = xla::Broadcast(one, {size});
+      if (has_weights) {
+        updates = weights;
+        zero = xla::Zero(ctx->builder(), dtype);
+      }
+    }
+
+    output = xla::Broadcast(zero, {output_shape.dimensions()});
+    
+    xla::XlaComputation assn_computation = [&] {
+      std::unique_ptr<xla::XlaBuilder> subb =
+      ctx->builder()->CreateSubBuilder("scatter_bincount");
+      xla::Shape param_shape = xla::ShapeUtil::MakeShape(dtype, {});
+      auto p0 = xla::Parameter(subb.get(), 0, param_shape, "p0");
+      auto p1 = xla::Parameter(subb.get(), 1, param_shape, "p1");
+      if (binary_output_) {
+        xla::One(subb.get(), dtype);
```

Comment:
Technically, I think this is not quite correct, as compiler might be allowed to give the initial value multiple times.

Can we use something like `And`?

Reply:
Can you expand on this?

Reply:
You're relying on the fact that the reducer will only be called with the initial value once.

While (I think) this is correct with the current implementation, I think it's not inconceivable to think of an implementation which will optimize `R(*A)` into `R(I, R(*A))` (adding another initial value upfront e.g. to help fusion), at which point returning 1 will not be correct.

Reply:
So do you want something like `xla::Add(xla::Zero, xla::One)`?

Reply:
I also see that in the update (like this one) it just return an Xla:op with an `Add`:
https://github.com/tensorflow/tensorflow/blob/2f8fc94a8cc1ce4d8e8a4ebe00c2545e76d8415e/tensorflow/compiler/tf2xla/kernels/scatter_nd_op.cc#L241-L251

Reply:
It seems the same case as `return y` in this XLA `TensorScatterUpdateOp`.  In any case I've added an explicit `xla::Or`